prometheus/collectors/dbstats_collector.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.15
+// +build go1.15
+
+package collectors
+
+import (
+	"database/sql"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type dbStatsCollector struct {
+	db *sql.DB
+
+	maxOpenConnections *prometheus.Desc
+
+	openConnections  *prometheus.Desc
+	inUseConnections *prometheus.Desc
+	idleConnections  *prometheus.Desc
+
+	waitCount         *prometheus.Desc
+	waitDuration      *prometheus.Desc
+	maxIdleClosed     *prometheus.Desc
+	maxIdleTimeClosed *prometheus.Desc
+	maxLifetimeClosed *prometheus.Desc
+}
+
+// DBStatsCollectorOpts defines the behavior of a db stats collector
+// created with NewDBStatsCollector.
+type DBStatsCollectorOpts struct {
+	// DriverName holds the name of driver.
+	// It will not used for empty strings.
+	DriverName string
```

Comment:
How about just calling this `Name` or `DBName`? A user could create different `*sql.DB` instances with the same driver.

In different news: Do you anticipate that we need more fields in this options struct? An options struct is great for later non-breaking additions, but if those additions never happen, it's just overhead. I cannot think about anything needed here right now, so I would tend towards just having a dbName argument on the NewDBStatsCollector constructor.